### PR TITLE
Remove KNOWN_CHROMECAST_INFO_KEY Dependency

### PR DESCRIPTION
**Pull request by icarome**
I've removed the dependency from the chromecast list of HA. Now the component it self create the chromecast list from network.
Not sure if the best way, but a working way until someone step up and fix properly.

### CODE REVIEW

**Comment by raman325** [comment]
Thanks for creating this! I was working on this too and pretty much chose the same approach as you. The only difference is that instead of doing a general discovery, I retrieve all the media players in the cast platform from the entity registry and use their unique IDs (the UUID of the device) to retrieve the devices. Not sure if it's a better approach than this, but I put my changes up in a branch in case you want to look at it: https://github.com/raman325/spotcast/tree/fix_2021_4

**Comment by hmmbob** [comment]
I'm trying to test this, but it errors on me:

```
2021-04-07 11:43:01 ERROR (SyncWorker_15) [custom_components.spotcast] Could not find device Tim from hass.data
2021-04-07 11:43:01 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall spotcast.start (c:9c4193b70307d736a58d59f3b0cca55a): account=t, entity_id=media_player.hmt, uri=spotify:track:5Oo<editted>0SozG, random_song=False, repeat=off, start_volume=101, force_playback=False, shuffle=False, offset=0>
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/core.py", line 1496, in catch_exceptions
await coro_or_task
File "/usr/src/homeassistant/homeassistant/core.py", line 1519, in _execute_service
await self._hass.async_add_executor_job(handler.job.target, service_call)
File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
result = self.fn(*self.args, **self.kwargs)
File "/config/custom_components/spotcast/__init__.py", line 298, in start_casting
spotify_cast_device = SpotifyCastDevice(
File "/config/custom_components/spotcast/__init__.py", line 415, in __init__
self.castDevice = self.getChromecastDevice(device_name)
File "/config/custom_components/spotcast/__init__.py", line 437, in getChromecastDevice
raise HomeAssistantError("Could not find device with name {}".format(device_name))
homeassistant.exceptions.HomeAssistantError: Could not find device with name Tim
```

The device is online and available in the HA Cast integration as well as the "entity overview" (with the entity_id media_player.hmt and showing friendly_name Tim)

**Comment by hmmbob** [comment]
I had a note here about testing failing on 2021.4.rc5, but when I rebooted to enable debug logging it all suddenly worked fine again? 🤷🏻 

Deleted that note, now confirming that this fix works fine here.

**Comment by FaserF** [comment]
It is working fine for me, but with this pull request I am getting the following info by HA:
`Detected integration that attempted to create another Zeroconf instance. Please use the shared Zeroconf via await homeassistant.components.zeroconf.async_get_instance(hass). Please report issue to the custom component author for spotcast using this method at custom_components/spotcast/sensor.py, line 46: known_devices, browser = pychromecast.get_chromecasts(zeroconf_instance=ChromeCastZeroconf.get_zeroconf())`

**Comment by hmmbob** [comment]
Interesting - not seeing that error over here (and I am (obviously) using cast and thus zeroconf).

What are you running? I am on Docker / HA Container.

**Comment by Aquilo77** [comment]
Hi, same error as FaserF but in helpers.py

Detected integration that attempted to create another Zeroconf instance. Please use the shared Zeroconf via await homeassistant.components.zeroconf.async_get_instance(hass). Please report issue to the custom component author for spotcast using this method at custom_components/spotcast/helpers.py, line 25: devices, browser = pychromecast.get_listed_chromecasts (uuids=uuids, zeroconf_instance=ChromeCastZeroconf.get_zeroconf()
 )

File: custom_components/spotcast/manifest.json [hunk 1]
```diff
@@ -5,8 +5,8 @@
   "requirements": [
     "spotify_token==1.0.0"
   ],
-  "homeassistant": "0.113.0",
-  "version": "3.4.7",
+  "homeassistant": "2021.4.0",
```
Comment (KapJI) [inline custom_components/spotcast/manifest.json:8]: In says nothing about "homeassistant" field:
https://developers.home-assistant.io/docs/creating_integration_manifest/

I think it should go to `hacs.json`.

File: custom_components/spotcast/sensor.py [hunk 2]
```diff
@@ -4,18 +4,17 @@
...
-from . import DOMAIN, KNOWN_CHROMECAST_INFO_KEY
-
+import homeassistant.components.zeroconf as zc
+import pychromecast
+from . import DOMAIN
+from homeassistant.components.cast.helpers import ChromeCastZeroconf
 _LOGGER = logging.getLogger(__name__)
-
-SENSOR_SCAN_INTERVAL_SECS = 10
+SENSOR_SCAN_INTERVAL_SECS = 30
```
Comment (KapJI) [inline custom_components/spotcast/sensor.py:12]: Why changing it?

File: custom_components/spotcast/__init__.py [hunk 3]
```diff
@@ -230,7 +226,17 @@ def play(client, spotify_device_id, uri, random_song, repeat, shuffle, position)
...
+            show_episodes_info = client.show_episodes(uri)
+            if show_episodes_info and len(show_episodes_info["items"]) > 0:
+                episode_uri = show_episodes_info["items"][0]["external_urls"]["spotify"]
+                _LOGGER.debug("Playing episode using uris (latest podcast playlist)= for uri: %s", episode_uri)
+                client.start_playback(device_id=spotify_device_id, uris=[episode_uri])         
+        elif uri.find("episode") > 0:
+            _LOGGER.debug("Playing episode using uris= for uri: %s", uri)
+            client.start_playback(device_id=spotify_device_id, uris=[uri])
+
+        elif uri.find("track") > 0:
```
Comment (KapJI) [inline custom_components/spotcast/__init__.py:239]: I think this should be in a separate PR.

File: custom_components/spotcast/__init__.py [hunk 4]
```diff
@@ -412,37 +418,17 @@ def __init__(self, hass, call_device_name, call_entity_id):
...
-                    for x in known_devices
-                    if known_devices[x].friendly_name == device_name
-                ),
-                None,
-            )
-
-        _LOGGER.debug("cast info: %s", cast_info)
+            cast_info = known_devices[0]
+        except IndexError:
+            cast_info = None
```
Comment (KapJI) [inline custom_components/spotcast/__init__.py:427]:
```suggestion
        cast_info = known_devices[0] if known_devices else None
```

File: custom_components/spotcast/__init__.py [hunk 5]
```diff
@@ -412,37 +418,17 @@ def __init__(self, hass, call_device_name, call_entity_id):
...
+        _LOGGER.info("cast info: %s", cast_info)
```
Comment (KapJI) [inline custom_components/spotcast/__init__.py:428]:
```suggestion
        _LOGGER.debug("cast info: %s", cast_info)
```

**Comment by FaserF** [comment]
@hmmbob using Homeassistant OS

**Comment by raman325** [comment]
> Hi, same error as FaserF but in helpers.py
> 
> Detected integration that attempted to create another Zeroconf instance. Please use the shared Zeroconf via await homeassistant.components.zeroconf.async_get_instance(hass). Please report issue to the custom component author for spotcast using this method at custom_components/spotcast/helpers.py, line 25: devices, browser = pychromecast.get_listed_chromecasts (uuids=uuids, zeroconf_instance=ChromeCastZeroconf.get_zeroconf()
> )

My branch implements these changes

**Comment by fondberg** [comment]
Thanks for helping out. Have you checked to see that this method in pychromecast is ok to run on the same process as another?
Earlier it was a problem and I had to go with the KNOWN_HOSTS option

**Comment by fondberg** [comment]
Anyways I'll merge and test this now

**Comment by fondberg** [comment]
I asked the core developer for pychromecast and they are not super happy with double scanning. They will add KNOWN_CHROMECAST_KEY back in the next release because this way will mean trouble for core.

I will do a release with this with a big disclaimer NOT to ask support for other cast issues that might arise

**Comment by hmmbob** [comment]
@fondberg I think @raman325 had a solution for this in his branch
https://github.com/raman325/spotcast/tree/fix_2021_4

**Comment by fondberg** [comment]
That is the same. I'm working one something. Gimme a few hours

**Comment by FaserF** [comment]
I am using @raman325 s branch since some days and it is working good so far.

this is the log message that I get:
`Logger: custom_components.spotcast.sensor
Source: custom_components/spotcast/sensor.py:22 
Integration: Start Spotify on chromecast (documentation) 
First occurred: 10 April 2021, 14:38:46 (1 occurrences) 
Last logged: 10 April 2021, 14:38:46

None`

**Comment by ssiergl** [comment]
related core roll back:

https://github.com/home-assistant/core/pull/49045

**Comment by fondberg** [comment]
Fixed together with core PRs. Help me upvote them (links in this PR) https://github.com/fondberg/spotcast/pull/176

File: custom_components/spotcast/__init__.py (hunk 3 above)
Comment (fondberg) [reply custom_components/spotcast/__init__.py:239]: I'll open a separate PR on this as it seems like good func

**Comment by frenck** [comment]
Upvotes don't make a difference.

**Comment by XalaTheShepard** [comment]
Many thanks for the update. The error is now gone, but I am so sorry to inform that it still does not work in HA OS. The Spotify playlist keeps on loading forever and no devices can be found. If I can contribute in any way, please let me know. 
image
